test/integration/setup/node-attestation/client.go

```diff
@@ -157,3 +167,206 @@ func doBanStep(ctx context.Context) {
 	}
 	// This doesn't return anything
 }
+
+// doX509popStep tests attestation using x509pop
+// Steps:
+// - Attest agent
+// - Renew agent
+// - Delete agent
+// - Reattest deleted agent
+// - Ban agent
+// - Reattest banned agent (must fails because it is banned)
+// - Delete agent
+// - Reattest deleted agent (must success after removing)
+func doX509popStep(ctx context.Context) {
+	c := itclient.New(ctx)
+	// Create an admin client to ban/delete agent
+	defer c.Release()
+	client := c.AgentClient()
+
+	// Attest agent
+	svidResp, err := x509popAttest(ctx)
+	if err != nil {
+		log.Fatalf("Failed to attest: %v", err)
+	}
+
+	// Renew agent
+	if err := x509popRenew(ctx, svidResp); err != nil {
+		log.Fatalf("failed to renew agent: %v", err)
+	}
+
+	// Delete agent
+	if err := deleteAgent(ctx, client, svidResp.Id); err != nil {
+		log.Fatalf("Failed to delete agent: %v", err)
+	}
+
+	// Reattest deleted agent
+	svidResp, err = x509popAttest(ctx)
+	if err != nil {
+		log.Fatalf("Failed to attest deleted agent: %v", err)
+	}
+
+	// Ban agent
+	if err := banAgent(ctx, client, svidResp.Id); err != nil {
+		log.Fatalf("Failed to ban agent")
+	}
+
+	// Reattest banned agent, it MUST fail
+	_, err = x509popAttest(ctx)
+	switch status.Code(err) {
+	case codes.OK:
+		log.Fatal("Error expected when attesting banned agent")
+	case codes.PermissionDenied:
+		if status.Convert(err).Message() != "failed to attest: agent is banned" {
+			log.Fatalf("Unnexpected error returned: %v", err)
+		}
+	default:
+		log.Fatalf("Unnexpected error returned: %v", err)
+	}
+
+	// Delete banned agent
+	if err := deleteAgent(ctx, client, svidResp.Id); err != nil {
+		log.Fatalf("Failed to delete agent: %v", err)
+	}
+
+	// Reattest deleted agent, now MUST be success
+	_, err = x509popAttest(ctx)
+	if err != nil {
+		log.Fatalf("Failed to attest deleted agent: %v", err)
+	}
+}
+
+// x509popAttest attest agent using x509pop
+func x509popAttest(ctx context.Context) (*types.X509SVID, error) {
+	log.Println("Attesting agent...")
+
+	// Create insecure connection
+	conn := itclient.NewInsecure(ctx)
+	defer conn.Release()
+	client := conn.AgentClient()
+
+	csr, err := x509.CreateCertificateRequest(rand.Reader, &x509.CertificateRequest{}, key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CSR: %v", err)
+	}
+
+	pair, err := tls.LoadX509KeyPair(*popCert, *popKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load key pair: %v", err)
+	}
+
+	data := &x509pop.AttestationData{
+		Certificates: pair.Certificate,
+	}
+	payload, err := json.Marshal(data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal payload: %v", err)
+	}
+
+	stream, err := client.AttestAgent(ctx)
```

Comment:
looks like CloseSend() is never called for this stream?

Reply:
it is expecged that stream is closed after we have a success challenge, I added another recv to verify that